text/0000-localizing-api-messages.md

```diff
@@ -0,0 +1,135 @@
+
+- Start Date: 03-29-2022
+- RFC PR: 
+- FOLIO Issue: 
+- Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
+- Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
+- Sub Group :
+
+# Localizing API (Backend) Messages
+
+## Summary
+  The purpose of this RFC is to define an approach that allows API developer to provide
+messages in the API response depending on the language preference set in the request. 
+
+
+## Motivation
+
+- Help users understand the information coming from the application backend using a language they understand better
+- To allow FOLIO platform to be used by users across the world 
+- To create a tremendous opportunity for growth that we can never could have achieved within just one country.
+
+### Terminology
+- Static Message - E.g. _This is something you need translate_
+- Static Message with placeholder - _You can ONLY renew {n} times_
+ 
+
+### In Scope Requirements/Use cases
+- Return localized messages based on the value passed in the accept-language header
+- Handle static messages, static messages with placeholders
+
+### Out of Scope Requirements/Use cases
+- [Plural Syntax](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Pluralsyntax)
+- Naming convention to be used for keys
+- Process for managing (decoupled from lokalise.com) translations
+- Returning formatted (HTML/Markdown) messages
+- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords) to break messages
+- Support customization per tenant
+- Process for back porting API/Backend messages similar to what we have for the front end 
+- Support for controlled vocabulary (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)
+
+## Detailed Explanation/Design
+- Locale specific translations MUST be stored in a properties file
+- Properties file naming convention
+  - default locale (en_US) - somename.properties
+  - specific locale - somename_fr.properties (For french),  somename_de.properties (For german)
+- Properties file MUST BE loaded using the classpath. AVOID using file locations
+- Properties file entries WILL use the standard key value pair format
+- When using application frameworks (spring, vertx, etc.) implement localization using the framework specific implementation (if there is one)
```

Comment:
I think it is important to acknowledge the trade off that we are introducing with this decision.

When this was discussed during the TC meeting, an impact of this was that the translation files could be in different places within the source code repository.

I understand that it makes it easier for developers, because folks are not working against the chosen tooling.

It creates a non-uniform interface to the outside world that needs to change the translations. That means we are putting the burden on the operations tooling and processes to accommodate this.

My bias tends to be the other way, we accept greater internal complexity (e.g. configuration of frameworks) to present less external complexity (a uniform location for translation property files).

I think this exposes our build (and localization) tooling to the potential of continual changes e.g. any time folks use a different language / framework (which some folks advocate for, we already have Vert.x, RMB, Spring Way, Grails, Node.js and Perl).

It undermines the idea that a module should comply with a set of expectations irrespective of the tooling choices.

(This is based upon the assumption that localization is one of those expectations, which I think is the intent of this proposal)

Reply:
@zburke 

> you're right, there's a "burden" to be carried here. I think placing that burden "on the operations tooling and processes" is exactly the right place for it. It consolidates this complexity (probably to a single script that syncs the values in our repos with the values in Lokalise or Smartling or Translized or whatever) instead of spreading it across every single repository. That seems like a great trade off to me.

I think making that decision and the trade offs of that approach explicit and prominent in this RFC would be great, to help both folks reviewing this now and folks using this as documentation in the future to understand this aspect.

I think that is important because this decision is broad in scope and impact, given that it affects how uniform the interface modules expose to the rest of the project are and has implications for build tooling and integration with services like Localize.

That could be even more important as FOLIO potentially moves towards a model where modules might be supported by a wider audience and are hosted elsewhere yet also need to integrate with these shared services (translation change management).

If I were to reflect my understanding I might express it something like the following.

Decision: each language / runtime MAY to use different file formats for static translations
Trade offs:
* allows each language / runtime variation to use more easily use existing tooling for localization
* Reduces the complexity of modules and effort for development teams as no mapping between external expectations and internal tool choices needs to be done
* Increases complexity of the builds which are involved in the centralised translation change management process (currently using Lokalise) and effort for DevOps folks

These trade offs assume that the tooling choice and structures will be standardised for each language / runtime rather than on a per module basis.

For that to be the case, we would need at a minimum, to define those standards for:
* Perl / Make
* Grails / Gradle
* Node.js / Yarn